index.js

```diff
@@ -70,23 +71,153 @@ function format(text, opts, addAlignmentSize) {
+function isSourceElement(node) {
+  if (node == null) {
+    return false;
+  }
+  switch (node.type) {
+    case "FunctionDeclaration":
+    case "BlockStatement":
+    case "BreakStatement":
+    case "ContinueStatement":
+    case "DebuggerStatement":
+    case "DoWhileStatement":
+    case "EmptyStatement":
+    case "ExpressionStatement":
+    case "ForInStatement":
+    case "ForStatement":
+    case "IfStatement":
+    case "LabeledStatement":
+    case "ReturnStatement":
+    case "SwitchStatement":
+    case "ThrowStatement":
+    case "TryStatement":
+    case "VariableDeclaration":
+    case "WhileStatement":
+    case "WithStatement":
+      return true;
+  }
+  return false;
+}
+
+function calculateRange(text, opts, ast) {
+  // Contract the range so that it has non-whitespace characters at its endpoints.
+  // This ensures we can format a range that doesn't end on a node.
+  const rangeStringOrig = text.slice(opts.rangeStart, opts.rangeEnd);
+  const startNonWhitespace = Math.max(
+    opts.rangeStart + rangeStringOrig.search(/\S/),
+    opts.rangeStart
+  );
+  let endNonWhitespace;
+  for (
+    endNonWhitespace = opts.rangeEnd;
+    endNonWhitespace > opts.rangeStart;
+    --endNonWhitespace
+  ) {
+    if (text[endNonWhitespace - 1].match(/\S/)) {
+      break;
+    }
+  }
+
+  const startNodeAndParents = findNodeAtOffset(ast, startNonWhitespace);
+  const endNodeAndParents = findNodeAtOffset(ast, endNonWhitespace);
+  const siblingAncestors = findSiblingAncestors(
+    startNodeAndParents,
+    endNodeAndParents
+  );
+  const startNode = siblingAncestors.startNode;
+  const endNode = siblingAncestors.endNode;
+  const rangeStart = Math.min(util.locStart(startNode), util.locStart(endNode));
+  const rangeEnd = Math.max(util.locEnd(startNode), util.locEnd(endNode));
+
+  const rangeString = text.slice(rangeStart, rangeEnd);
+
+  // Try to extend the range backwards to the beginning of the line.
+  // This is so we can detect indentation correctly and restore it.
   // Use `Math.min` since `lastIndexOf` returns 0 when `rangeStart` is 0
-  const rangeStart = Math.min(
-    opts.rangeStart,
-    text.lastIndexOf("\n", opts.rangeStart) + 1
+  const rangeStart2 = Math.min(
+    rangeStart,
+    text.lastIndexOf("\n", rangeStart) + 1
   );
-  // Use `text.length - 1` as the maximum since `indexOf` returns -1 if `fromIndex >= text.length`
-  const fromIndex = Math.min(opts.rangeEnd, text.length - 1);
-  const nextNewLineIndex = text.indexOf("\n", fromIndex);
-  const rangeEnd = (nextNewLineIndex < 0 ? fromIndex : nextNewLineIndex) + 1; // Add one to make rangeEnd exclusive
-
-  if (0 < rangeStart || rangeEnd < text.length) {
-    const rangeString = text.substring(rangeStart, rangeEnd);
-    const alignmentSize = getAlignmentSize(
-      rangeString.slice(0, rangeString.search(/[^ \t]/)),
-      opts.tabWidth
-    );
+  const indentString = text.slice(rangeStart2, rangeStart);
+  if (indentString.trim() == "") {
```

Comment:
Sure, fixed in https://github.com/prettier/prettier/pull/1659/commits/29e247c0f2736053ae433ba5e38ed7b5026dfbc6. As you can see in the commit, this also fixed the indentation when formatting just the `IfStatement` in the following code:

```js
try {
1;if (condition) {
    body
  }
}
catch (err) {}
```